IO: sketch out alternative callback based API

ifreund opened the pull request:
The control flow of callbacks is much more explicit and easier to
understand than that of async. Furthermore, async is one of the least
stable areas of Zig currently and this is not expected to change in the
near future.

This is a draft as I have not yet implemented all operations supported by io_uring and need to add more tests to ensure the implementation is working properly.

Furthermore the `callconv(.C)` required for safe type erasure of the function pointers requires passing the error union results by constant pointer which is rather ugly. Perhaps I can think of a different way to do this though without requiring the user of the API to preform unsafe casts.

jorangreef commented:
Thanks @ifreund !

The new linked list names (`in`/`out`/`push`) are a nice improvement, as well as swapping the null check around for an if unwrap, and encapsulating head/tail into the FIFO struct to simplify the reset just before iterating the list. Nice work! Good to see you've also got the same assertions in all the key places. Thanks for getting us started on proper error sets too!

I have some thoughts on the API for the callsite:

If we work backwards from [Journal.write_sectors()](https://github.com/coilhq/tigerbeetle/blob/beta/src/vr.zig#L812-L842), which will do the equivalent of a `pwriteAll()`, how would we want to manage our memory in Journal when submitting IO? If we start there then perhaps we can further reduce the interface for our new callback methods to be limited to only what we need, without having to be too generic, and so get away from `callconv(.C)`?

Along this line, I've been thinking we should aim for a single struct allocation per concurrent write op, to replace the `appending_frame` we have on the Replica instance, so that we then only have to manage memory for one thing, one basket. We would then merge the Completion and Context structs into this "one thing", a flat struct, which would also store the result, and which knows where to hand off control flow when the callback fires, instead of the Completion pointing to the Context and the Context memory being managed separately.

This "one thing, one basket" is then less flexible, and it's more of a straight-jacket compared to allowing two separate Completion and Context structs, with the Completion struct simply pointing to the Context, but I think we could use it to simplify the call site, because this would also allow the IO abstraction to hide stuff like completions and type args away from the call site.

For example, at present, with our generic IO submission method:

```
var fd: os.fd_t = 0;
var openat: IO.Completion = undefined;
io.submit(&openat, *os.fd_t, &fd, .{
    .openat = .{
        .fd = linux.AT_FDCWD,
        .path = path,
        .flags = os.O_CLOEXEC | os.O_RDWR | os.O_CREAT,
        .mode = 0o666,
        .callback = openat_callback,
    },
});
```

This requires the callsite to know about completions and contexts, and manage memory for both. It also has to pass the type of the context.

If we were to do a syscall specific function (were you thinking still to do this?), and if we started with:

`io.write(fd: os.fd_t, buffer: []const u8, offset: u64) !usize`

...what's the bare minimum we would need to add to this? For example, we could add `context` and `callback`, but what if we could just add `completion` (at the front) and let `callback` be under that?

At present, we also pass `context` and `result` to the callback function. This style is almost taken for granted when using async callbacks. But what if we didn't do that? We only need the callback to tell us when the IO completes, that's its main purpose and we could use it only for that, storing the result rather in the completion struct and then making the callback a method on the completion struct. This might pave the way for callbacks to different syscalls to all have the same function signature, and perhaps get away from callconv(.C)?

ifreund commented:
I think I've found a way to hide the `callconv(.C)` and some of the other ugliness in the abstraction, giving a nicer function signature of (for example):
```zig
    pub fn openat(
        self: *IO,
        completion: *Completion,
        comptime Context: type,
        context: Context,
        comptime callback: fn (result: OpenatError!void, context: Context) void,
        fd: os.fd_t,
        path: [*:0]const u8,
        flags: u32,
        mode: os.mode_t,
    ) void;
```
This requires making the callback function comptime known, but I don't expect this to be restricting in practice. 

I don't think merging the `Completion` and `Context` structs would really help here. The test case doesn't really demonstrate what would be considered typical usage of the callback API. I expect `Context` to generally be a pointer to whatever struct is responsible for preforming the request (i.e. `MessageBus` or `Replica`).

There is however another reasonable way I see to provide access to some context within the callback, though it would require embedding the completion inside the context struct:
```zig
const MessageBus = struct {
    // some fields....
    completion: Completion,
    // more fields and stuff...
};

fn callback(completion: *Completion, result: OpenatError!os.fd_t) void {
    const message_bus = @fieldParentPtr(*MessageBus, "completion", completion);
    // handle result of the openat 
}
```
This would lose some flexibility in how the user of the API can manage the memory of the `Context` and `Completion` structs but it would simply the code and allow for uniform callback function signatures.

ifreund commented:
> Along this line, I've been thinking we should aim for a single struct allocation per concurrent write op, to replace the appending_frame we have on the Replica instance, so that we then only have to manage memory for one thing, one basket. We would then merge the Completion and Context structs into this "one thing", a flat struct, which would also store the result, and which knows where to hand off control flow when the callback fires, instead of the Completion pointing to the Context and the Context memory being managed separately.

With this as the goal I'm inclined to purse the `@fieldParentPtr()` based API I proposed which I think fits well into this model.

jorangreef commented:
> I think I've found a way to hide the callconv(.C) and some of the other ugliness in the abstraction, giving a nicer function signature of (for example):

```zig
pub fn openat(
    self: *IO,
    completion: *Completion,
    comptime Context: type,
    context: Context,
    comptime callback: fn (result: OpenatError!void, context: Context) void,
    fd: os.fd_t,
    path: [*:0]const u8,
    flags: u32,
    mode: os.mode_t,
) void;
```

Nice!

To see if I understand, would the above then be able to translate into something like this:

```zig
const Journal = struct {
    // some fields....
    io: *IO,
    completion_write: Completion,
    // more fields and stuff...

    fn write(self: *Journal, buffer: []const u8, offset: u64) void {
        self.io.write(
            &self.completion_write,
            Journal,
            self,
            write_callback,
            self.fd,
            buffer,
            offset,
        );
    }
    
    fn write_callback(self: *Journal, completion: *Completion, result: WriteError!usize) void {
        
    }
};
```

That's based on your `openat` signature before the `fieldParentPtr` trick which would require the callback to be outside the Journal. I think if we can pass `self: *Journal` to the callback that would be better.

Have I understood correctly?

> This requires making the callback function comptime known, but I don't expect this to be restricting in practice.

100% agree.

ifreund commented:
Yep, your code snippet is a good example of how I picture this being used in most cases. We could probably use `anytype` to avoid the need to explicitly pass the type of the `Context` pointer and instead infer it from the pointer passed. The tradeoff here is that the explicitness can help readability and thus `anytype` is usually avoided when possible even if it could be used to avoid explicitly passing a type (e.g. see `std.mem.eql()`).

I'll get my proposed revision to the callback API implemented shortly.

ifreund commented:
Implemented the revised version for accept/close/openat. Next step is implementing the rest of the operations we need and porting the tests from the current `io.zig`

jorangreef commented:
> We could probably use anytype to avoid the need to explicitly pass the type of the Context pointer and instead infer the type for the callback's signature from the pointer passed. The tradeoff here is that the explicitness can help readability and thus anytype is usually avoided when possible even if it could be used to avoid explicitly passing a type (e.g. see std.mem.eql()).

Agreed re: not using `anytype`. `std.mem.eql()` is a great example.

> Implemented the revised version for accept/close/openat.

Looks awesome!

I'll leave a few small ideas directly on the source.

One last big change for the API though, where we now have:

```
self.io.write(
    &self.completion_write,
    Journal,
    self,
    write_callback,
    self.fd,
    buffer,
    offset,
);
```

What do you think about tweaking the order to:

```
self.io.write(
    Journal,
    self,
    &self.completion_write,
    write_callback,
    self.fd,
    buffer,
    offset,
);
```

This puts the context first, ahead of the completion, because the contexts (MessageBus, Connection, Journal) will be bigger than the completions, owning them and interpreting them. Also puts the comptime type first like in `std.mem.eql`. I guess it depends on whether we see the context as the biggest thing that makes sense of the result, or whether we see it only as something related to the callback function.

Beyond that we could also put the callback last, but it probably belongs just after the completion, then we're keeping all the callback IO args together.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+
+pub const IO = struct {
+    /// An intrusive first in/first out linked list.
+    const Fifo = struct {
```
Let's do `FIFO` here instead of `Fifo` in keeping with `IO` vs `Io`.

ifreund replied:
Happy to change that for consistency, I've been trying to follow the style of the project so far even though it conflicts with the official zig style guide on this point and on the naming of functions (the zig style guide and standard library use camelCase instead of snake_case).

As for the precise algorithm used to name types this is what has been settled on (though it could almost certainly be better documented and is not yet followed everywhere in the standard library) https://github.com/ziglang/zig/issues/1097#issuecomment-722112034.

I personally try to waste as little time as possible on coding style, all I want is consistency, so I'm happy to go with whatever you prefer here.

jorangreef replied:
Great, yes I was following that issue too, and had hoped all "the ayes would have it" in favor of snake_case functions but alas! It's also a pretty big change for Zig to make on behalf of everyone if it had gone through. I noticed that you also stuck to all the current styles of the project, thanks.

I'm happy that we keep the snake_case method names within the project, because that extra separator between words helps readability, especially in our VR implementation so far where we have methods like `find_latest_header_break_between`... it's the closest one can get to spaces but without using them. That was also something I always liked about C and Ruby.

> I personally try to waste as little time as possible on coding style, all I want is consistency, so I'm happy to go with whatever you prefer here.

Yes, same here. That's also why I find camelCase difficult, nobody knows whether it's XmlHttpRequest or XMLHTTPRequest etc. as that issue highlighted.

But in this case I actually really just thought FIFO here because it's an abbreviation! 😄

ifreund replied:
> Yes, same here. That's also why I find camelCase difficult, nobody knows whether it's XmlHttpRequest or XMLHTTPRequest etc. as that issue highlighted.

The algorithm described in the comment I linked resolves this ambiguity, but point taken. I'll rename to FIFO.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+
+    ring: IO_Uring,
+
+    /// Completions not yet submitted to the kernel and waiting on available space in the
```
Operations instead of Completions? A completion is more like a CQE, and this is not an SQE either because we're not yet in the queue.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+    }
+
+    /// Pass all queued submissions to the kernel and run the event loop
+    /// until there is no longer any i/o pending.
```
Let's do `I/O` instead of `i/o`.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+    /// until there is no longer any i/o pending.
+    pub fn run(self: *IO) !void {
+        while (self.queued + self.submitted > 0 or
+            self.unqueued.out != null or self.completed.out != null)
```
If this was Haiku then it would be cool to reuse the completed queue to handle a full submission queue, then we can make this a one-liner, but I think the way you've got it is the simplest already.

ifreund replied:
Thinking about this a bit more I don't think we actually need the `completed` FIFO since we are using callbacks not async. Do you see a reason that would stop us from getting rid of it and just executing the callbacks directly in `flush_completions()`?

jorangreef replied:
Great, I was hoping this would just be food for thought. Let me take a look and see...

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+                completion.run();
+            }
+            // Again, run on a copy of the list to avoid an infinite loop
+            out = self.unqueued.out;
```
Let's do these two loops in their own blocks and not reuse `out`.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+        context: ?*c_void,
+        callback: fn (context: ?*c_void, completion: *Completion, result: *const c_void) callconv(.C) void,
+
+        fn prep(completion: *Completion) void {
+            completion.io.get_sqe(completion);
+        }
+
+        fn complete_prep(completion: *Completion, sqe: *io_uring_sqe) void {
```
Maybe `prep()` should be `queue()` or `enqueue()` and `complete_prep` should be `prep` because that's what they're doing?

ifreund replied:
Yeah that makes sense, as you may have noticed I struggled a bit with naming things here :)

jorangreef replied:
No worries, I actually think you improved most of the names, FIFO, in/out/push etc.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+            sqe.user_data = @ptrToInt(completion);
+        }
+
+        fn run(completion: *Completion) void {
```
Instead of `run()` what do you think of `complete()` or `execute_callback()`, seeing as we already use `run` in the event loop sense for `IO` itself, and the completion here has finished running?

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+        fn run(completion: *Completion) void {
+            switch (completion.operation) {
+                .accept => |op| {
+                    const result = result: {
```
The labelled break expression here is fine, but is it possible to reduce this slightly by using an if expression directly instead?

ifreund replied:
Already changed that to a less verbose version in the lasted commit. I personally am not a huge fan of 
```zig
const x = if (foo) blk: {
    // ...
} else blk: {
    // ...
};
```
as I think having a single block is more readable so I tend to opt for the single block version when the if isn't a single statement.

jorangreef replied:
> I personally am not a huge fan

Yeah, actually me too. I also prefer a single block. Happy to do what you think here.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+        }
+    };
+
+    /// This union encodes the set of operations supported as well as their arguments.
+    const Operation = union(enum) {
+        accept: struct {
+            socket: os.fd_t,
+            address: os.sockaddr,
+            address_size: os.socklen_t,
+            flags: u32,
+        },
+        close: struct {
+            fd: os.fd_t,
+        },
+        openat: struct {
+            fd: os.fd_t,
+            path: [*:0]const u8,
+            flags: u32,
+            mode: os.mode_t,
+        },
+    };
+
```
This could almost be a code generator or DSL! Such a terse description of everything supported!

ifreund replied:
This is pretty similar to what my [wayland scanner](https://github.com/ifreund/zig-wayland) creates from protocol xml. Zig is lovely for stuff like this :)

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,463 @@
+        completion.* = .{
+            .io = self,
+            .context = context,
+            .callback = struct {
+                fn wrapper(ctx: ?*c_void, comp: *Completion, res: *const c_void) callconv(.C) void {
+                    callback(
+                        @intToPtr(Context, @ptrToInt(ctx)),
+                        comp,
+                        @intToPtr(*const AcceptError!os.socket_t, @ptrToInt(res)).*,
+                    );
+                }
+            }.wrapper,
```
Neat trick!

ifreund replied:
Thanks! We could also take this a bit further and use `@call(.{ .modifier = .always_inline }, ... )` to call the callback here though I don't have any performance numbers to back it up so I left it off for now.

jorangreef replied:
Good to know! We can always profile later and this is all internal within the API.

ifreund commented:
> This puts the context first, ahead of the completion, because the contexts (MessageBus, Connection, Journal) will be bigger than the completions, owning them and interpreting them. Also puts the comptime type first like in `std.mem.eql`. I guess it depends on whether we see the context as the biggest thing that makes sense of the result, or whether we see it only as something related to the callback function.

Done, I agree with your reasoning.

Note there is currently one small difference from your code snippet, namely that the Context type is the pointer type itself not the child type:
```zig
self.io.write(
    *Journal,
    self,
    &self.completion_write,
    write_callback,
    self.fd,
    buffer,
    offset,
);
```
While we could interpret `Context` as a pointer child type and pass `*Context` to the callback this would make the API slightly less flexible as optional or other kinds of pointers would not be permitted. I don't really have a use case for them though. It's not a major change either way but would be better to do now before we have a bunch of code using this API that would need to be updated.

jorangreef commented:
> While we could interpret Context as a pointer child type and pass *Context to the callback this would make the API slightly less flexible as optional or other kinds of pointers would not be permitted. I don't really have a use case for them though. It's not a major change either way but would be better to do now before we have a bunch of code using this API that would need to be updated.

Writing `Journal` as the `Context` type instead of `*Journal` was just an accident on my part. Let's keep it as you have it!

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,902 @@
+                        os.EIO => error.InputOutput,
+                        os.ENOSPC => error.NoSpaceLeft,
+                        else => |errno| os.unexpectedErrno(@intCast(usize, errno)),
+                    } else assert(completion.result == 0);
```
What is the value of `const result =` on line 219 if we only have an `assert()` on `completion.result` here on line 226 without providing anything explicitly (i.e. as we would normally do on line 215)?

ifreund replied:
I realize now that this probably isn't as clear as it was in my head. The return type of `assert()` is `void` so result (which is of type `CloseError!void`) would be assigned the value `void` in this case to indicate that there was no error.

ifreund replied:
It would be more explicit to do
```zig
blk: {
    assert(completion.result == 0);
    break :blk {};
};
```

jorangreef replied:
> I realize now that this probably isn't as clear as it was in my head. The return type of assert() is void so result (which is of type CloseError!void) would be assigned the value void in this case to indicate that there was no error.

No, you're right. I had the 0/void type mixed up. Let's leave it as is, or as you prefer.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,902 @@
+        fn complete(completion: *Completion) void {
+            switch (completion.operation) {
+                .accept => {
+                    const result = if (completion.result < 0) switch (-completion.result) {
```
This is clean. What is your feeling on this if expression? If you prefer the block that's also fine.

ifreund replied:
Since it's just a single statement I don't think the block is necessary.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,902 @@
+    pub const Completion = struct {
+        io: *IO,
+        result: i32 = undefined,
+        next: ?*Completion = undefined,
```
I think we should set the default value for `next` to `null` as soon as possible since we know that will be the initial value and we're not waiting for the value to be defined later. This would also make line 140 less of an issue if it was accidentally left out. We could then rather assert on line 140: `assert(completion.next == null)`.

ifreund replied:
Setting the default value to `null` wouldn't be sufficient to avoid setting null explicitly on line 140 as completions may be re-used without re-initialization in some cases if an operation fails with EINTR.

What we could do here is unconditionally set `next` to null in `FIFO.push()`

jorangreef replied:
I think it's good that `FIFO.push()` asserts `next` is `null`, rather than unconditionally setting it, because that would help surface this kind of reuse without correctly setting `next` back to `null`, and we can then make sure that all our calls to `FIFO.push()` are already getting it right.

Let's set the default to `null` because that's a valid initial state, and then also make sure we reset `next` back to `null` as soon as the `next` pointer is no longer valid, which is the moment when it's shifted out a queue. `next` should either be `null` or a valid pointer if it's in a FIFO queue, but as soon as it's no longer in a queue it should be set back to `null`. So this is still an unconditional reset, but at the moment it's shifted rather than pushed.

ifreund replied:
Implemented in the latest commit.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,902 @@
+                },
+                .close => {
+                    const result = if (completion.result < 0) switch (-completion.result) {
+                        os.EINTR => {}, // A success, see https://github.com/ziglang/zig/issues/2425.
```
Should we do `0` here instead of the empty block to be more explicit? I think `{}` will implicitly be `0` (I think that's how Zig does `void` or `u0`?) but perhaps we should go directly to `0`?

ifreund replied:
`{}` (an empty block) has type `void` in zig, which is the type we want on a success for result which is a `CloseError!void` error union. Using `0` here wouldn't compile.

jorangreef replied:
That's perfect then. I don't know where I got the notion that `close()` should have a result of 0 instead of void!

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,902 @@
+                },
+            },
+        };
+        completion.io.enqueue(completion);
```
We could also do these as `self.enqueue(completion)` since we are within `*IO`.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,902 @@
+}
+
+test "" {
+    std.testing.refAllDecls(IO);
```
Learned something new!

ifreund commented:
Implemented some of the cleanup you suggested and added tests for the currently supported operations.

Last TODO on this branch is support for timeouts.

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,1189 @@
+            // Loop on a copy of the linked list, having reset the linked list first, so that any
+            // synchronous append on running a completion is executed only the next time round
+            // the event loop, without creating an infinite suspend/resume cycle.
+            {
+                var copy = self.completed;
+                self.completed = .{};
+                while (copy.pop()) |completion| completion.complete();
+            }
+            // Again, loop on a copy of the list to avoid an infinite loop
+            {
+                var copy = self.unqueued;
+                self.unqueued = .{};
+                while (copy.pop()) |completion| self.enqueue(completion);
+            }
```
Really nice!

ifreund replied:
Thanks :)

jorangreef commented on src/io_callbacks.zig:
```diff
@@ -0,0 +1,1189 @@
+                    linux.io_uring_prep_send(sqe, op.socket, op.buffer, op.flags);
+                },
+                .timeout => |op| {
+                    // TODO Use proper 64 bit timespec type: https://github.com/ziglang/zig/pull/8118
```
We are tracking master so we could do this one now.

ifreund replied:
Sure, I been using a rather out of date master commit so that `vr.zig` still builds, the `std.fmt` changes to require `{s}` for printing slices as a string broke things.

jorangreef replied:
Ah, yes I thought you would have spotted something like that since I hadn't updated for awhile. I've updated all the format specifiers in `vr.zig` so that it now builds on `zig-0.8.0-dev.1471+4e9894cfc`.

jorangreef commented:
Great work.
